### ISSUE TITLE
App fails on launch (lb-tsc)

### ISSUE THREAD

**Issue by JaredTheGreater** [strongloop/loopback4-example-shopping#459]
Hello,

I've been having difficulty launching this example app. I'm using an Ubuntu server and following the guide as closely as I can. I admit I'm not an experienced NodeJS dev. Please find the error and environment details below. Any assistance is very much appreciated:

Node: v12.14.0
NPM: 6.13.4
Command: sudo npm start
Error:
> loopback4-example-shopping-monorepo@1.1.1 prestart /home/jared/node/loopback/loopback4-example-shopping
> npm run build
> loopback4-example-shopping-monorepo@1.1.1 build /home/jared/node/loopback/loopback4-example-shopping
> lerna run build
> loopback4-example-recommender@1.1.1 build /home/jared/node/loopback/loopback4-example-shopping/packages/recommender
> lb-tsc
> loopback4-example-shopping@1.1.1 build /home/jared/node/loopback/loopback4-example-shopping/packages/shopping
> lb-tsc
src/__tests__/integration/recommender-grpc.integration.ts(7,43): error TS2307: Cannot find module 'grpc'.
src/__tests__/integration/recommender-rest.integration.ts(8,26): error TS2307: Cannot find module '@loopback/http-server'.
src/index.ts(8,15): error TS2307: Cannot find module '@loopback/http-server'.
src/recommendation-grpc.ts(14,8): error TS2307: Cannot find module 'grpc'.
src/recommendation-grpc.ts(16,24): error TS2307: Cannot find module '@grpc/proto-loader'.
src/recommendation-grpc.ts(50,5): error TS7006: Parameter 'call' implicitly has an 'any' type.
src/recommendation-grpc.ts(51,5): error TS7006: Parameter 'callback' implicitly has an 'any' type.
src/recommendation-rest.ts(8,26): error TS2307: Cannot find module '@loopback/http-server'.
npm ERR! code ELIFECYCLE
npm ERR! errno 2
npm ERR! loopback4-example-recommender@1.1.1 build: `lb-tsc`
npm ERR! Exit status 2
npm ERR! 
npm ERR! Failed at the loopback4-example-recommender@1.1.1 build script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.
npm WARN Local package.json exists, but node_modules missing, did you mean to install?
npm ERR! A complete log of this run can be found in:
npm ERR!     /home/jared/.npm/_logs/2019-12-30T17_44_55_386Z-debug.log
npm ERR! code ELIFECYCLE
npm ERR! errno 2
npm ERR! loopback4-example-shopping-monorepo@1.1.1 build: `lerna run build`
npm ERR! Exit status 2
npm ERR! 
npm ERR! Failed at the loopback4-example-shopping-monorepo@1.1.1 build script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.
npm ERR! A complete log of this run can be found in:
npm ERR!     /home/jared/.npm/_logs/2019-12-30T17_44_55_437Z-debug.log
npm ERR! code ELIFECYCLE
npm ERR! errno 2
npm ERR! loopback4-example-shopping-monorepo@1.1.1 prestart: `npm run build`
npm ERR! Exit status 2
npm ERR! 
npm ERR! Failed at the loopback4-example-shopping-monorepo@1.1.1 prestart script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.
npm ERR! A complete log of this run can be found in:
npm ERR!     /home/jared/.npm/_logs/2019-12-30T17_44_55_453Z-debug.log

**Comment by hacksparrow** [comment]
Run `npm i` in the project directory. Also, you will need to have MongoDB and Redis instances running.

**Comment by JaredTheGreater** [comment]
Thanks, @hacksparrow for your quick response. I did run the `npm i` command as you suggested and received the following:
```
> husky@3.1.0 install /home/jared/node/loopback/loopback4-example-shopping/node_modules/husky
> node husky install

husky > Setting up git hooks
EACCES: permission denied, open '../../.git/hooks/applypatch-msg'
husky > Failed to install

> core-js@2.6.11 postinstall /home/jared/node/loopback/loopback4-example-shopping/node_modules/core-js
> node -e "try{require('./postinstall')}catch(e){}"

Thank you for using core-js ( https://github.com/zloirock/core-js ) for polyfilling JavaScript standard library!

The project needs your help! Please consider supporting of core-js on Open Collective or Patreon: 
> https://opencollective.com/core-js 
> https://www.patreon.com/zloirock 

Also, the author of core-js ( https://github.com/zloirock ) is looking for a good job -)


> husky@3.1.0 postinstall /home/jared/node/loopback/loopback4-example-shopping/node_modules/husky
> opencollective-postinstall || exit 0

Thank you for using husky!
If you rely on this package, please consider supporting our open collective:
> https://opencollective.com/husky/donate

npm WARN lifecycle loopback4-example-shopping-monorepo@1.1.1~postinstall: cannot run in wd loopback4-example-shopping-monorepo@1.1.1 lerna bootstrap (wd=/home/jared/node/loopback/loopback4-example-shopping)
added 1222 packages from 1409 contributors and audited 44885 packages in 20.298s

15 packages are looking for funding
  run `npm fund` for details

found 0 vulnerabilities
```
The npm warning looks important, but I don't really understand it (especially the "wd" argument, it's a valid directory). Any insights? Is this why I'm not able to solve the issue?

I also ensured I had MongoDB installed (which I didn't apparently as it installed quite a bit) and tested:
```
● mongodb.service - An object/document-oriented database
   Loaded: loaded (/lib/systemd/system/mongodb.service; enabled; vendor preset: enabled)
   Active: active (running) since Mon 2019-12-30 18:09:16 UTC; 4min 38s ago
     Docs: man:mongod(1)
 Main PID: 18703 (mongod)
    Tasks: 23 (limit: 2227)
   CGroup: /system.slice/mongodb.service
           └─18703 /usr/bin/mongod --unixSocketPrefix=/run/mongodb --config /etc/mongodb.conf

Dec 30 18:09:16 api systemd[1]: Started An object/document-oriented database.
```
test:
```
jared@api:~/node/loopback/loopback4-example-shopping$ mongo --eval 'db.runCommand({ connectionStatus: 1 })'
MongoDB shell version v3.6.3
connecting to: mongodb://127.0.0.1:27017
MongoDB server version: 3.6.3
{
        "authInfo" : {
                "authenticatedUsers" : [ ],
                "authenticatedUserRoles" : [ ]
        },
        "ok" : 1
}
```
Additionally, I ensured redis-server was installed and running (which it was):
```
● redis-server.service - Advanced key-value store
   Loaded: loaded (/lib/systemd/system/redis-server.service; enabled; vendor preset: enabled)
   Active: active (running) since Mon 2019-12-30 18:11:27 UTC; 12min ago
     Docs: http://redis.io/documentation,
           man:redis-server(1)
  Process: 19825 ExecStop=/bin/kill -s TERM $MAINPID (code=exited, status=0/SUCCESS)
  Process: 19828 ExecStart=/usr/bin/redis-server /etc/redis/redis.conf (code=exited, status=0/SUCCESS)
 Main PID: 19850 (redis-server)
    Tasks: 4 (limit: 2227)
   CGroup: /system.slice/redis-server.service
           └─19850 /usr/bin/redis-server 127.0.0.1:6379

Dec 30 18:11:27 api systemd[1]: Starting Advanced key-value store...
Dec 30 18:11:27 api systemd[1]: redis-server.service: Can't open PID file /var/run/redis/redis-server.pid (yet?) after start: No such file or di
Dec 30 18:11:27 api systemd[1]: Started Advanced key-value store.
```
test:
```
127.0.0.1:6379> ping
PONG
127.0.0.1:6379> set test "It's working!"
OK
127.0.0.1:6379> get test
"It's working!"
```
Thanks again for your assistance.

**Comment by JaredTheGreater** [comment]
Hello @hacksparrow. I believe I might have resolved this issue. It seems there are notorious issues with running npm with the `sudo` command. I googled the npm warning type and learned about an `npm install` flag called `--unsafe-perm` that will assist with this npm issue (https://stackoverflow.com/questions/18136746/npm-install-failed-with-cannot-run-in-wd). When I ran the `sudo npm install --unsafe-perm` command, I was then able to successfully launch the app. Thank you for helping me brainstorm. Afterall, I would have had to install MongoDB and your advice saved me that headache.
